hnn_core/tests/test_io.py

```diff
@@ -0,0 +1,134 @@
+import os.path as op
+from numpy.testing import assert_allclose
+from h5io import write_hdf5
+import pytest
+
+import hnn_core
+from hnn_core import read_network
+from hnn_core import simulate_dipole
+from hnn_core import read_params
+from hnn_core import jones_2009_model, law_2021_model, calcium_model
+
+hnn_core_root = op.dirname(hnn_core.__file__)
+params_fname = op.join(hnn_core_root, 'param', 'default.json')
+
+
+@pytest.mark.parametrize("network_model",
+                         [law_2021_model, calcium_model,
+                          jones_2009_model])
+def test_network_io(tmp_path, network_model):
+    # For simulation to be shorter(Discuss)
+    params = op.join(hnn_core_root, 'param', 'default.json')
+    if isinstance(params, str):
+        params = read_params(params)
+    params['N_pyr_x'] = 3
+    params['N_pyr_y'] = 3
+    params['celsius'] = 37.0
+    params['threshold'] = 0.0
+
+    # Instantiating network along with drives
+    net = network_model(params=params, add_drives_from_params=True)
+
+    # Adding bias
+    net.add_tonic_bias(cell_type='L2_pyramidal', amplitude=1.0)
+
+    # Test __eq__ method
+    net_copy = net.copy()
+    assert net_copy == net
+
+    # Adding electrode arrays
+    electrode_pos = (1, 2, 3)
+    net.add_electrode_array('el1', electrode_pos)
+    electrode_pos = [(1, 2, 3), (-1, -2, -3)]
+    net.add_electrode_array('arr1', electrode_pos)
+
+    # Writing network
+    net.write(tmp_path / 'net.hdf5')
+
+    # Testing when overwrite is False and same filename is used
+    with pytest.raises(FileExistsError,
+                       match="File already exists at path "):
+        net.write(tmp_path / 'net.hdf5', overwrite=False)
+
+    # Reading network
+    net_read = read_network(tmp_path / 'net.hdf5')
+    assert net == net_read
+
+    # Simulating network
+    dpls1 = simulate_dipole(net, tstop=2, n_trials=1, dt=0.5)
+    dpls2 = simulate_dipole(net_read, tstop=2, n_trials=1, dt=0.5)
+    for dpl1, dpl2 in zip(dpls1, dpls2):
+        assert_allclose(dpl1.times, dpl2.times, rtol=0.00051, atol=0)
+        for dpl_key in dpl1.data.keys():
+            assert_allclose(dpl1.data[dpl_key],
+                            dpl2.data[dpl_key], rtol=0.000051, atol=0)
+    # Writing simulated network and reading it
+    net.write(tmp_path / 'net_sim.hdf5')
+    net_sim = read_network(tmp_path / 'net_sim.hdf5')
+    assert net == net_sim
+
+    # For cell response vsec isec bug (Not checked by __eq__)
```

Comment:
I'm not sure I remember, what is the bug?

Reply:
Its not a bug exactly. From what I remember `vsec` and `isec` were not getting compared in the `__eq__` method and putting this logic there was breaking some flow. I will look into it again and figure out a way to place in the `__eq__` method as that is the correct place for such comparison.

Reply:
Moved this to __eq__ in cell response